bokehjs/src/coffee/renderer/geo/tile_layer.coffee

```diff
@@ -0,0 +1,416 @@
+
+  meters_to_tile: (mx, my, level) ->
+    [px, py]= @meters_to_pixels(mx, my, level)
+    return @pixels_to_tile(px, py)
+
+  get_tile_meter_bounds: (tx, ty, level) ->
+    # expects tms styles coordinates (bottom-left origin)
+    [xmin, ymin] = @pixels_to_meters(tx * @tile_size, ty * @tile_size, level)
+    [xmax, ymax] = @pixels_to_meters((tx + 1) * @tile_size, (ty + 1) * @tile_size, level)
+    return [xmin, ymin, xmax, ymax]
+
+  get_tile_geographic_bounds: (tx, ty, level) ->
+    bounds = @get_tile_meter_bounds(tx, ty, level)
+    [minLon, minLat, maxLon, maxLat] = @utils.meters_extent_to_geographic(bounds)
+    return [minLon, minLat, maxLon, maxLat]
+
+  get_tiles_by_extent: (extent, level, tile_border=2) ->
+
+    # unpack extent and convert to tile coordinates
+    [xmin, ymin, xmax, ymax] = extent
+    [txmin, tymin] = @meters_to_tile(xmin, ymin, level)
+    [txmax, tymax] = @meters_to_tile(xmax, ymax, level)
+
+    # add tiles which border
+    txmin -= tile_border
+    tymin -= tile_border
+    txmax += tile_border
+    tymax += tile_border
+
+    tiles = []
+    for ty in [tymax..tymin] by -1
+      for tx in [txmin..txmax] by 1
+        tiles.push([tx, ty, level, @get_tile_meter_bounds(tx, ty, level)])
+
+    tiles = @sort_tiles_from_center(tiles, [txmin, tymin, txmax, tymax])
+    return tiles
+
+class TMSTileProvider extends MercatorTileProvider
+
+  get_image_url: (x, y, z) ->
+    return @url.replace("{X}", x).replace('{Y}', y).replace("{Z}", z)
+
+class WMTSTileProvider extends MercatorTileProvider
+
+  get_image_url: (x, y, z) ->
+    [x, y, z] = @tms_to_wmts(x, y, z)
+    return @url.replace("{X}", x).replace('{Y}', y).replace("{Z}", z)
+
+class QUADKEYTileProvider extends MercatorTileProvider
+
+  quadkey_to_tile_xyz: (quadKey) ->
+    '''
+    Computes tile x, y and z values based on quadKey.
+    '''
+    tileX = 0
+    tileY = 0
+    tileZ = quadKey.length
+    for i in [tileZ...0] by -1
+
+      value = quadKey.charAt(tileZ - i)
+      mask = 1 << (i - 1)
+
+      if value == '0'
+        continue
+
+      else if value == '1'
+        tileX |= mask
+
+      else if value == '2'
+        tileY |= mask
+
+      else if value == '3'
+        tileX |= mask
+        tileY |= mask
+
+      else
+        throw new TypeError("Invalid Quadkey: " + quadKey)
+
+    return [tileX, tileY, tileZ]
+
+  tile_xyz_to_quadkey: (x, y, z) ->
+    '''
+    Computes quadkey value based on tile x, y and z values.
+    '''
+    quadKey = ''
+    for i in [z...0] by -1
+      digit = 0
+      mask = 1 << (i - 1)
+      if(x & mask) != 0
+        digit += 1
+      if(y & mask) != 0
+        digit += 2
+      quadKey += digit.toString()
+    return quadKey
+
+  get_image_url: (x, y, z) ->
+    [x, y, z] = @tms_to_wmts(x, y, z)
+    quadKey = @tile_xyz_to_quadkey(x, y, z)
+    return @url.replace("{QUADKEY}", quadKey)
+
+
+
+class TileLayerView extends Glyph.View
+
+  trace: (msg) ->
+    console.warn "TileLayerView :: " + msg.toString()
```

Comment:
We will disallow omitting parentheses in function calls soon, besides some special cases like `require "module name"`, so make sure to use parentheses consequently. Also, in this case you may want to use string interpolation, i.e. `console.warn("TileLayerView :: #{msg}")`. Also, we have a bokehjs-wide logger, so you you should use it instead (see `bokehjs/src/coffee/common/logger`).